### ISSUE TITLE
A Task returning method being called in a Void method

### ISSUE THREAD

**Issue by PaymonK** [Geeksltd/GCop#30]
If a Task returning method is called in a void method, warn to say:

"Change this method to return a Task"

**Comment by mohsad06** [comment]
@PaymonK Could you please provide an example for this?

**Comment by PaymonK** [comment]
`void SomeMethod()
{
     someObject.GetMethodThatReturnsTask();
}`

In the above example, SomeMethod() can return without GetMethodThatReturnsTask() is actually completed.

There should be a warning on SomeMethod to say it should return a Task instead of being void.

**Comment by Foroughi** [comment]
@reporter: FYI its gonna be in Cop.Extra

**Comment by PaymonK** [comment]
Why GCop.Extra?

**Comment by Foroughi** [comment]
@PaymonK  in the given example the method `GetMethodThatReturnsTask` needs to be checked if it returns a System.Threading.Task or not. this method can be implemented anywhere in the project. which means I need to have access to a higher level than current class.

**Comment by PaymonK** [comment]
But can we not find the type of a locally invoked expression in Roslyn?!

**Comment by Foroughi** [comment]
@PaymonK , it depends on the definition of word "local". if by local you mean the types which are implemented in the current parsing file, the answer is yes. if by local you mean the types which are defined in the current project, the answer is no. I should raise this point that as far as i understand (and I'm almost sure) in the current context you will have access to symbols (and types) who : 

1. are implemented in current parsing file
2. are referenced manually into current compilation context (check [here](https://github.com/dotnet/roslyn/issues/25745#event-1544958477))
3. Net Framework standard classes
4- Referenced assembelies into current compilation context

Also I ran a test on it. I made a simple project. containing two classes : 

```
    public class TestClass1
    {       
        public void func1()
        {
            TestClass2 obj = new TestClass2();
            obj.func2();
        }
    } 

    public class TestClass2
    {
        
        public Task func2()
        {
            return Task.Run(()=>{});
        } 

    }     
```

These classes are in different files, same project. Assuming Im parsing `TestClass1`, Still cannt access to the symbol or type of `.func2()` 

![screenshot 70](https://user-images.githubusercontent.com/3727486/38811659-563e689c-419f-11e8-8353-1c565efb5d18.png)

**Comment by Karvan** [comment]
@PaymonK I think we need to add this rule to the GCop.Extra extension.

**Comment by PaymonK** [comment]
That's fine

**Comment by Foroughi** [comment]
@reporter : Added on [NuGetPackages/9647ea6bcf156c8b2a7debb148a71fa8e7ac0f86](https://gitlab.com/Paymon/GCop/tree/9647ea6bcf156c8b2a7debb148a71fa8e7ac0f86)

**Comment by Foroughi** [comment]
Here are some details of the new rule : 

## General Details:

- ID : **655**
- Category : **Refactoring**
- Message : **Change this method to return a Task**
- Severity: **Warning**
- Package : **GCop.Extra**

## Technical Details : 

The Rule will look for a method with the following specs :

- Returning value should be `void`
- The method should contain a `MemberAccessSyntax` which is returning a `Task`


## Test Result :
I used the two classes I mentioned in previous comments in this task


In case of any changes needs to be done on this rule please let me know.
